development/code-guidelines/tasks-metatasks-guidelines.md

```diff
@@ -0,0 +1,138 @@
+```java
+org.osc.core.broker.service.tasks.k8s.securitygroup.MyTask  // Tasks or MetaTasks specific to security group and k8s
+```
+```java
+org.osc.core.broker.service.tasks.distributedappliance.MyTask  // Tasks or MetaTasks specific to distributed appliance, agnostic to virtualization platform and integration point
+```
+```java
+org.osc.core.broker.service.tasks.ost.sdncontroller.securitygroup.MyTask  // Tasks or MetaTasks specific to SDN controller, OpenStack and security group
+```
+
+**DON'T:**
+```java
+org.osc.core.broker.service.tasks.securitymanager.MyTask // Mandatory entity name part missing.
+```  
+```java
+org.osc.core.broker.service.tasks.domain.SomeDomainTask // Domain is an entity specific to security manager this package is likely missing "securitymanager".
+```
+```java
+org.osc.core.broker.service.tasks.distributedappliances.MyTask // Do NOT use PLURAL in the entity name part.
+```
+```java
+org.osc.core.broker.service.tasks.securitymanager.device.member.MyTask // There should be only one part for the entity, this should be "devicemember" instead.
+```
+```java
+org.osc.core.broker.service.tasks.k8s.pod.create.MyTask // The package name should ALWAYS end with the entity name, do NOT create more granular packages like "*.create".
+```
+```java
+org.osc.core.broker.service.tasks.conform.securitygroup.MyTask // "conform" is not a part name.
+```
+
+####  Class Names  
+
+Tasks and MetaTasks class names must follow the pattern `[ACTION][OPT:VIRTUALIZATION_PLATFORM][OPT:INTEGRATION_POINT][ENTITY_NAME][OPT: ENTITY_ACTION_DETAILS][TASK_METATATASK]`.  
+* `ACTION`:  This is the action performed by the task or metatask with respect to the `ENTITY` part. Common actions are CRUD calls to integration points, checking the state of some entity, conforming some entity with multiple integration points and virtualization platform, etc. Expected names are: `Create`, `Upsert`, `Update`, `Delete`, `Conform`, `Check`, `CreateOrUpdate`, `DeleteOrUpdate`, etc.   
+>  **Note:** For consistency DO NOT use other semantically equivalent names like `Register` (equivalent to `Create` or `Upsert`) or `Remove` (equivalent to `Delete`) .  Stick to the terms mentioned above. If you come accross a new action and are sure that nothing similar already exists in the code you may create a new action name.  
+>  **Note:** Observe there is a difference between the action `Upsert` and `CreateOrUpdate`. The former indicates the task will invoke some API that creates or updates and entity (if it already exists) on the database or integration point. The latter should be used only for metatasks and it indicates that the metatask will conditionally add a `Create` or `Update` task to the execution graph.  
+* `VIRTUALIZATION_PLATFORM`: The possible values for this *optional* part are `Ost` or `K8s`. This part indicates that the task or metatask is specific to a given virtualization platform. Only tasks and metatasks that are platform agnostic should not have this part.  
```

Comment:
I see some definition like VIRTUALIZATION_PLATFORM is repeated, rather we could have a definition section and refer to that section.